### ISSUE TITLE
Dropping packets when using whitelist

### ISSUE THREAD

**Issue by jhart-r7** [zmap/zmap#560]
Recently I've been using zmap to scan a specific set of targets.  zmap has two modes of doing this currently, `-I` (list of IPs) and `-w` (CIDRs to whitelist).  I'm launching zmap as follows:

```
zmap -r 0 -B 400M  -s 465 -p 465 -b /path/to/blacklist.conf --seed 1234  -M tcp_synscan -O csv -f timestamp_ts,saddr,sport,daddr,dport,ipid,ttl -w /tmp/input.txt --output-filter="success = 1 && repeat = 0 && sport = 465" --shards 4 --shard 0
```

In this case, the input file has ~2M assets that I suspect have 465/TCP open.

I've noticed that when using `-w` my zmap instances drop a large number of packets right at the beginning:

```
Jun 11 21:43:38.718 [WARN] monitor: Dropped 119611 packets in the last second, (119656 total dropped (pcap: 119656 + iface: 0))
```

It seems partially related to the scan rate, because as I drop the scan rate from our default of 400MB/s down to 100MB/s, the drops disappear.  Similarly, if I keep the 400MB/s rate and scan all IPv4, no drops occur.  Even more strange, if I switch to `-I`, no drops occur.

For now I think I can just switch to `-I` to work around the issue however that then means targets must be specified as individual IPs rather than CIDRs .  

Any ideas what might be causing this?

**Comment by dadrian** [comment]
First off, apologizes about the over 1 year delay.

I added a section in the Wiki about this, at the bottom of https://github.com/zmap/zmap/wiki/Block-and-Allow-Lists. After talking with @zakird and looking at the code, we have a couple of thoughts:

* The packet drop is on the receive side (pcap stats)
* There's nothing that inherently connects `-I` or `-w` to recv performance
* In general, for disparate hosts, `-I` should be more CPU-efficient than `-w`, since it will be a constant lookup into a bitmap, rather than a cache-trashing lookup into a tree dependent on the size of the allowlist.

Depending on the makeup of your allowlist, a couple things could be happening:
* If the allowlist can be written as a small number of large CIDRs, it might be more performant than `-I`, leading to increased scan rate, leading to increased hit rate, leading to drop
* If you have a low-end system where receive and send are competing for resources, and `-I` is more efficient, then it is likely allowing receive to have enough CPU time to actually receive the packets.

In either case, since you do have an allowlist, you likely have a much higher hit rate than a normal Internet-wide scan, which means the receive thread has to do more work per packet sent by the send thread. It's generally a good idea to slow down scans using allowlists.

**tl;dr** I don't think there's a bug here, and this is part of tuning speeds and results

**Comment by zhangyoufu** [comment]
Output processing and pcap recv share the same thread. If you have slow I/O or you pipe stdout to some other program, the recv thread may stuck and unable to process incoming packets.

```
(gdb) bt
#0  0x00007f51cdf9721f in write () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007f51cdf1800d in _IO_file_write () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007f51cdf19ad1 in _IO_do_write () from /lib/x86_64-linux-gnu/libc.so.6
#3  0x00007f51cdf175f8 in _IO_file_sync () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007f51cdf0b546 in fflush () from /lib/x86_64-linux-gnu/libc.so.6
#5  0x000055ed7cff2ac8 in csv_process ()
#6  0x000055ed7cfe31b3 in handle_packet ()
#7  0x000055ed7d0171e5 in pcap_handle_packet_mmap ()
#8  0x000055ed7d017654 in pcap_read_linux_mmap_v3 ()
#9  0x000055ed7cff3742 in recv_packets ()
#10 0x000055ed7cfe3505 in recv_run ()
#11 0x000055ed7cfe5e94 in start_recv ()
#12 0x00007f51ce081609 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#13 0x00007f51cdfa8293 in clone () from /lib/x86_64-linux-gnu/libc.so.6
```

**Comment by dadrian** [comment]
This is the point of the ztee utility included with ZMap. It will buffer an arbitrary amount of output in memory, in order to compensate for a slower downstream service.
